atl-checker/src/algorithms/certain_zero/search_strategy/linear_optimize.rs

```diff
@@ -0,0 +1,505 @@
+use crate::algorithms::certain_zero::search_strategy::linear_optimize::Ranges::Range;
+use crate::algorithms::certain_zero::search_strategy::{SearchStrategy, SearchStrategyBuilder};
+use crate::atl::Phi;
+use crate::edg::atlcgsedg::AtlVertex;
+use crate::edg::Edge;
+use crate::game_structure::lcgs::ast::{BinaryOpKind, DeclKind, ExprKind, Identifier};
+use crate::game_structure::lcgs::ir::intermediate::{IntermediateLcgs, State};
+use crate::game_structure::lcgs::ir::symbol_table::SymbolIdentifier;
+use minilp::OptimizationDirection::{Maximize, Minimize};
+use minilp::{ComparisonOp, OptimizationDirection, Problem};
+use priority_queue::PriorityQueue;
+use std::cmp;
+use std::collections::HashMap;
+use std::option::Option::Some;
+use std::sync::Arc;
+use BinaryOpKind::{
+    Addition, And, Division, Equality, GreaterOrEqual, GreaterThan, Implication, Inequality,
+    LessOrEqual, LessThan, Multiplication, Or, Subtraction, Xor,
+};
+
+/// A SearchStrategyBuilder for building the LinearOptimizeSearch strategy.
+pub struct LinearOptimizeSearchBuilder {
+    pub game: IntermediateLcgs,
+}
+
+impl SearchStrategyBuilder<AtlVertex, LinearOptimizeSearch> for LinearOptimizeSearchBuilder {
+    fn build(&self) -> LinearOptimizeSearch {
+        LinearOptimizeSearch::new(self.game.clone())
+    }
+}
+
+/// Search strategy using ideas from linear programming to order next vertices,
+/// based on distance from the vertex to a region that borders the line between true/false in the formula.
+pub struct LinearOptimizeSearch {
+    /// Priority based on distance, lowest distance highest priority
+    queue: PriorityQueue<Edge<AtlVertex>, i32>,
+    game: IntermediateLcgs,
+    /// Maps the hash of a Phi and usize of state, to a result distance
+    result_cache: HashMap<(Arc<Phi>, usize), i32>,
+    /// Maps usize from proposition, to a hashmap mapping symbols to the ranges they need to be within,
+    /// to satisfy the proposition
+    proposition_cache: HashMap<usize, HashMap<SymbolIdentifier, Ranges>>,
```

Comment:
We have typealiases for `usize` called `State` and `Proposition` which you might want to use.